### PR TITLE
feat: uberschrift for Vue

### PR DESCRIPTION
**Pull request by randowha**
Despite installing the esbuild-plugin-vue and adding a default config for tsup, the building process for the vue package does not work yet

### CODE REVIEW

**Review by moritzjacobs — CHANGES_REQUESTED** [review]
Thanks for working on this! I'm looking forward to it!

I just did a quick check and also couldn't make the build work. Need to dedicate some more time later.

A few notes:
- there are lint errors all over the place, your vscode should warn you and / or autofix them when you save
- some more todos in comments:

**Review by jannikkeye — COMMENTED** [review]
I've fixed the build process, I tried using tsup but that didn't want to work nicely with the vue plugins. Since we already have Vite installed I opted to use Vite library mode, which works well.

Also streamlined the eslint config usage so that there is a dedicated react and vue config. Fixed the errors along the way. The vue config currently doesn't include any vue rules, I've tested this before but ran into problems with rules.

We should revisit that.

@randowha sorry for hijacking but I thought this would be most efficient and I think its a lot to put fixing the build process and repo setup things on you as well.

File: packages/vue/src/uberschrift.test.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	});
+
+	[1, 2, 3, 4, 5, 6, 7, 8].forEach((i) => {
+		test(`level ${i}`, () => {
+			const TestComponent = defineComponent({
+				setup() {
+					return {
+						...useHx(),
+					};
+				},
```
[the remark this answers is not in the document]
Comment (jannikkeye) [reply packages/vue/src/uberschrift.test.ts]: I've addressed these, see my comment.

File: packages/vue/src/demo.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { createApp } from "vue";
+import TestCase from "./TestCase.vue";
+// import "./index.css";
```
[the remark this answers is not in the document]
Comment (jannikkeye) [reply packages/vue/src/demo.ts]: I've addressed this, see my comment.

File: apps/vue-vite-example/.gitignore
Comment (moritzjacobs) [inline apps/vue-vite-example/.gitignore:1]: I think this file can be removed, the main .gitignore in the root folder should take care of it.

File: apps/vue-vite-example/index.html
Comment (moritzjacobs) [inline apps/vue-vite-example/index.html:1]: A few of the files need formatting, you can run `npm run format` in the root folder to make prettier do its thing.

File: apps/vue-vite-example/package.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	"private": true,
+	"version": "0.0.0",
+	"type": "module",
+	"scripts": {
+		"dev": "vite",
+		"build": "vue-tsc && vite build",
+		"preview": "vite preview"
+	},
+	"dependencies": {
+		"uberschrift": "*",
```
Comment (moritzjacobs) [inline apps/vue-vite-example/package.json:12]: Shouldn't this be `uberschrift-vue`? 🤔

File: apps/vue-vite-example/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    /* Linting */
+    "strict": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "noFallthroughCasesInSwitch": true
+  },
+  "include": ["src/**/*.ts", "src/**/*.tsx", "src/**/*.vue"],
+  "references": [{ "path": "./tsconfig.node.json" }]
+}
```
Comment (moritzjacobs) [inline apps/vue-vite-example/tsconfig.json:25]:
```suggestion
{
	"extends": "@uberschrift/tsconfig/base.json",
	"include": ["src/**/*.ts", "src/**/*.tsx", "src/**/*.vue"],
	"references": [{ "path": "./tsconfig.node.json" }]
}

```

… if I'm not mistaken.

Comment (moritzjacobs) [reply apps/vue-vite-example/tsconfig.json:25]: Not sure what we need the references here? @jannikkeye ?

File: apps/vue-vite-example/tsconfig.node.json
Comment (moritzjacobs) [inline apps/vue-vite-example/tsconfig.node.json:1]: see above

File: packages/react/vitest.config.ts.timestamp-1713343108310-08cdd84b3bb9b.mjs
Comment (moritzjacobs) [inline packages/react/vitest.config.ts.timestamp-1713343108310-08cdd84b3bb9b.mjs:1]: ?

File: packages/vue/src/uberschrift.test.ts
Comment (moritzjacobs) [inline packages/vue/src/uberschrift.test.ts:1]: - [ ] We need to remember to change this filename when we land on a package name.

File: packages/vue/src/uberschriftSymbol.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export const uberschriftSymbol = Symbol("level");
```
Comment (moritzjacobs) [inline packages/vue/src/uberschriftSymbol.ts:1]: A wild `Symbol` appears! It' super effective!

**Review by moritzjacobs — CHANGES_REQUESTED** [review]
Niiiiice! It works!
A few nitpicks, but I love it! ❤️
Also we need to talk about the TODOs I already commented.

More TODOs for when we're ready to merge:
- [ ] A generous round of dependency updates
- [ ] We need to overhaul [the documentation page](https://uberschrift.peerigon.io/) to include it
- [ ] We should replace the README symlink in the root with a framework agnostic README

File: packages/vue/src/uberschrift.test.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import { describe, expect, test } from "vitest";
+import { render } from "@testing-library/vue";
+import TestCase from "./TestCase.vue";
+import { useHx } from "./useHx";
```
[the remark this answers is not in the document]
Comment (aubcel) [reply packages/vue/src/uberschrift.test.ts]: https://vuejs.org/guide/reusability/composables.html#naming
Vue actually copied react here speaking of naming conventions, but compsables are NOT Hooks: https://vuejs.org/guide/reusability/composables.html#vs-react-hooks

File: packages/vue/package.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
+{
+	"name": "uberschrift-vue",
```
[the remark this answers is not in the document]
Comment (aubcel) [reply packages/vue/package.json:2]: `v-uberschrift`?
so it is in the pattern of vue directives which always start with `v-...`

[the remark this answers is not in the document]
Comment (moritzjacobs) [reply packages/vue/package.json:2]: Doesn't that convey that it's a directive, which it is not?

[the remark this answers is not in the document]
Comment (aubcel) [reply packages/vue/package.json:2]: yeah, fair enough 😬

File: packages/vue/src/Hx.vue [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+const props = withDefaults(
+	defineProps<{
+		increment?: number;
+	}>(),
+	{
+		increment: 0,
+	},
+);
+
+const previousLevel = inject(uberschriftLevel, 1);
```
Comment (aubcel) [inline packages/vue/src/Hx.vue:14]: basically i only changed the naming to have a descriptive variable naming.

File: packages/vue/src/uberschriftSymbol.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import { InjectionKey } from "vue";
+
+export const uberschriftSymbol = Symbol("uberschriftLevel");
+
+export const uberschriftLevel = uberschriftSymbol as InjectionKey<number>;
```
Comment (aubcel) [inline packages/vue/src/uberschriftSymbol.ts:5]: This way we have a neat typing in the inject and provide part without having to type them again and again.

File: packages/vue/src/useHx.ts [hunk 10]
```diff
@@ -1,10 +1,10 @@
 import { inject } from "vue";
-import { uberschriftSymbol } from "./uberschriftSymbol";
+import { uberschriftLevel } from "./uberschriftSymbol";
 
 type HeadingTagName = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
 
 export const useHx = () => {
-	const level = inject<number>(uberschriftSymbol) ?? 1;
```
Comment (aubcel) [inline packages/vue/src/useHx.ts]: oh wow, i didn't expect that to fail in the tests. Ususally the as the second parameter is a default in case nothing was provided... :/
Now I see why it was like that before 🤷🏻‍♀️ 
Will revert the change

**Review by moritzjacobs — CHANGES_REQUESTED** [review]
Works for me, please review #10, to get the last few things out of the way. We will work on a changeset based release overhaul later today.
